internal/memory/pool.go

```diff
@@ -0,0 +1,323 @@
+//
+// For efficiency and to reduce fragmentation an entire region is allocated when requested.
+// When there's no available region of requested size, we try to allocate more memory for
+// this particular size by splitting the next available larger region into smaller ones, e.g.
+// if there's no region available for size class 0, we try splitting a region from class 1,
+// then class 2 etc, until we are able to do so or hit the upper limit.
+type PoolAllocator struct {
+	pools [memoryClassNumber]*memoryPool
+}
+
+var _ MappedRegion = &region{}
+var _ Allocator = &PoolAllocator{}
+
+func (r *region) Offset() uint64 {
+	return r.offset
+}
+
+func (r *region) Size() uint64 {
+	sz, _ := GetMemoryClassSize(r.class)
+	return sz
+}
+
+func newEmptyMemoryPool() *memoryPool {
+	return &memoryPool{
+		free: make(map[uint64]*region),
+		busy: make(map[uint64]*region),
+	}
+}
+
+func NewPoolMemoryAllocator() PoolAllocator {
+	pa := PoolAllocator{}
+	p := newEmptyMemoryPool()
+	// by default we allocate a single region with maximum possible size (class type)
+	p.free[0] = &region{
+		class:  memoryClassNumber - 1,
+		offset: 0,
+	}
+	pa.pools[memoryClassNumber-1] = p
+	return pa
+}
+
+// FindFreeOffset finds free memory region with minimum offset that can fit a given memory class.
+func (pa *PoolAllocator) FindFreeOffset(memCls classType) (uint64, error) {
+	if memCls >= memoryClassNumber {
+		return 0, ErrInvalidMemoryClass
+	}
+
+	for i := memCls; i < memoryClassNumber; i++ {
+		pi := pa.pools[i]
+		if pi == nil || len(pi.free) == 0 {
+			continue
+		}
+
+		// start off with max offset
+		target := 4 * GigaByte
+		for offset := range pi.free {
+			if offset < target {
+				target = offset
+			}
+		}
+		return target, nil
+	}
+	return 0, ErrNotEnoughSpace
+}
+
+// Allocate checks memory pool of class `memCls` and returns a free region with minimal offset, if none available,
+// tries to expanding current `memCls` pool.
+//
+// Internally it's done via moving a region from free pool into a busy pool
+func (pa *PoolAllocator) Allocate(memCls classType) (MappedRegion, error) {
+	if memCls >= memoryClassNumber {
+		return nil, ErrInvalidMemoryClass
+	}
+
+	// Find smallest available offset
+	offset, err := pa.FindFreeOffset(memCls)
+	if err != nil {
+		return nil, err
+	}
+
+	// Mark region at a given offset as busy
+	if r, err := pa.markBusy(memCls, offset); err != nil {
+		if err := pa.Expand(memCls); err != nil {
+			if err == ErrInvalidMemoryClass {
+				return nil, ErrNotEnoughSpace
+			}
+			return nil, err
+		}
+	} else {
+		return r, nil
+	}
+
+	// Now the region at `offset` location should be available
+	if r, _ := pa.markBusy(memCls, offset); r != nil {
+		return r, nil
+	}
+
+	return nil, ErrNotEnoughSpace
+}
+
+// Expand tries to recursively split a bigger memory region into smaller ones until it succeeds or hits the upper limit
+func (pa *PoolAllocator) Expand(clsType classType) error {
```

Comment:
I feel split is probably a better name for this function. What do you think?

Reply:
not sure... we are splitting larger memory blocks into small ones, but the actual request is: "can you create more regions of `clsType` by splitting a larger one?", not sure what's the best word to describe this, but I welcome inputs 😄